### PR TITLE
Integrate action dropdown button

### PR DESCRIPTION
**Pull request by johnsonw**

### CODE REVIEW

File: static/js/lib/format-bytes.js [hunk 1]
```diff
@@ -3,10 +3,10 @@
 // Use of this source code is governed by a MIT-style
 // license that can be found in the LICENSE file.
 
-export default function (bytes, precision) {
-  const units = ['B', 'kiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB'];
+export default function(bytes, precision) {
```
Comment (jgrund) [inline static/js/lib/format-bytes.js]: Can we do this in a separate patch?

File: static/js/lib/format-big-number.js [hunk 2]
```diff
@@ -3,12 +3,12 @@
 // Use of this source code is governed by a MIT-style
 // license that can be found in the LICENSE file.
 
-export default function (num, precision, strict = false) {
-  const units = ['', 'k', 'M', 'B', 'T'];
+export default function(num, precision, strict = false) {
```
Comment (jgrund) [inline static/js/lib/format-big-number.js]: Can we do this in a separate patch?

**Comment by johnsonw** [comment]
Action buttons on the filesystem detailed page in a disabled state while the filesystem is being created:
image

File: static/js/api.js [hunk 3]
```diff
@@ -431,6 +431,12 @@ var Api = function() {
     });
   };
 
+  var callPromise = function(verb, url, api_args) {
+    return new Promise((res, rej) => {
+      call(verb, url, api_args, x => res(x), e => rej(e));
+    });
+  }
+
```
Comment (jgrund) [inline static/js/api.js]: How about:

```js
const callPromise = (verb, url, api_args) =>
    new Promise((res, rej) => {
      call(verb, url, api_args, x => res(x), e => rej(e));
    });
```

File: static/js/api.js [hunk 4]
```diff
@@ -450,6 +456,7 @@ var Api = function() {
   return {
     enable: enable,
     call : call,
+    callPromise: callPromise,
```
Comment (jgrund) [inline static/js/api.js]: You could omit the value here.

File: static/js/services/generate_command_dropdown.js [hunk 5]
```diff
@@ -1,96 +1,137 @@
...
+
+  if (locks) fn(locks);
+};
+
+const addDestroyer = (uuid, fn) => {
+  destroyers.set(uuid, fn);
+};
+
+export const removeListener = uuid => {
+  console.log("removing listener for uuid ", uuid);
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Make sure to remove this before landing.

File: static/js/services/generate_command_dropdown.js [hunk 6]
```diff
@@ -1,96 +1,137 @@
...
+  listeners.delete(uuid);
+}
+
+export const removeDestroyer = uuid => {
+  console.log("removing destroyer for uuid", uuid);
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Make sure to remove this before landing.

File: static/js/services/generate_command_dropdown.js [hunk 7]
```diff
@@ -1,96 +1,137 @@
...
+const onMessage = ({data}) => {
+  locks = JSON.parse(data);
+
+  listeners.forEach(fn => {
+    fn(locks);
+  });
+};
+
+const removeAllListeners = () => {
+  console.log("removing listeners: ", listeners.length);
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Make sure to remove this before landing.

File: static/js/services/generate_command_dropdown.js [hunk 8]
```diff
@@ -1,96 +1,137 @@
...
+  window.removeEventListener("message", onMessage);
+  listeners.clear();
+};
+
+const destroyAllButtons = () => {
+  console.log("destroying all buttons");
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Make sure to remove this before landing.

File: static/js/services/generate_command_dropdown.js [hunk 9]
```diff
@@ -1,96 +1,137 @@
...
+  destroyers.forEach(fn => fn());
+  destroyers.clear();
+}
+
+// Handle receiving locks from the parent
+window.addEventListener("message", onMessage);
+
+// Remove the event listeners when the page unloads
+window.addEventListener('unload', () => {
+  console.log('destroying event listeners');
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Make sure to remove this before landing.

File: static/js/services/generate_command_dropdown.js [hunk 10]
```diff
@@ -1,96 +1,137 @@
...
+let id = () => {};
+
+// Called whenever the table:
+// A. Finishes rendering
+// B. A pagination table update completed
+export function dataTableInfoCallback (settings) {
+  settings.aoData.forEach(({nTr: row, _aData: data}) => {
+    let cell = Array.from(row.cells).find(cell => cell.classList.contains("actions-cell"));
+    if (cell != null) {
+      const div = document.createElement('div');
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Is this guaranteed to be called once? Do you want to check if a div exists already when this fn is entered?

File: static/js/services/generate_command_dropdown.js [hunk 11]
```diff
@@ -1,96 +1,137 @@
...
+  const instance = init({
+    uuid,
+    records: [record],
+    locks: {},
+    flag: undefined,
+    tooltip_placement: placement,
+    tooltip_size: undefined
+  });
+
+  console.log("adding listener with uuid", uuid);
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Make sure to remove this before landing.

File: static/js/services/generate_command_dropdown.js [hunk 12]
```diff
@@ -1,96 +1,137 @@
...
+  addListener(uuid, locks => {
+    console.log('setting locks');
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Make sure to remove this before landing.

File: static/js/services/generate_command_dropdown.js [hunk 13]
```diff
@@ -1,96 +1,137 @@
...
+    instance.set_locks(locks);
+  });
+
+  addDestroyer(uuid,  () => {
+    console.log("destroying and freeing", uuid);
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Make sure to remove this before landing.

File: static/js/services/generate_command_dropdown.js [hunk 14]
```diff
@@ -1,96 +1,137 @@
...
+    .forEach(row => {
+      const cell = Array.from(row.cells)
+        .find(cell => cell.classList.contains("actions-cell"));
+      
+      if (cell) {
+        const div = cell.querySelector("div");
+        if (div) {
+          const uuid = div.id;
+          if (uuid) {
+            window.generateCommandDropdown.removeListener(uuid);
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Why does this need to be `window`?

File: static/js/services/generate_command_dropdown.js [hunk 15]
```diff
@@ -1,96 +1,137 @@
...
+            const destroyer = destroyers.get(uuid);
+            if (destroyer)
+              destroyer();
+              
+            window.generateCommandDropdown.removeDestroyer(uuid);
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Why does this need to be `window`?

File: static/js/target_detail.js [hunk 16]
```diff
@@ -18,8 +18,7 @@ var TargetDetail = Backbone.View.extend({
     $(this.el).find('.ui-dialog-content').html(rendered);
     $(this.el).find('.tabs').tabs({'show': function(event, ui) {view.tab_select(event, ui)}});
 
-    var generateCommandDropdown = angular.element('html').injector().get('generateCommandDropdown');
-    generateCommandDropdown.generateDropdown($(this.el).find('div[command-dropdown]'), cleanModel);
+    window.generateCommandDropdown.generateDropdown($(this.el).find('div[command-dropdown]')[0], cleanModel);
```
Comment (jgrund) [inline static/js/target_detail.js]: Where does cleanup happen?

File: templates/base.html [hunk 17]
```diff
@@ -1,190 +1,394 @@
 <!DOCTYPE html>
```
Comment (jgrund) [inline templates/base.html]: There are a ton of formatting changes in this file, can we just include the ones needed for this patch?

File: templates/filesystem_detail.html [hunk 18]
```diff
@@ -142,12 +141,15 @@
...
         bServerSide: true,
-        fnRowCallback: generateCommandDropdown.dataTableCallback(6, _.identity, _.isEmpty),
+        fnInfoCallback: window.generateCommandDropdown.dataTableInfoCallback,
         fnServerData: function (url, data, callback, settings) {
           var tableInstance = this;
 
           var callbacks = {
-            success: callback,
+            success: (...args) => {
+              window.generateCommandDropdown.cleanupButtons(tableInstance[0]);
```
Comment (jgrund) [inline templates/filesystem_detail.html]: When does this get called?

File: templates/mgt_tab.html [hunk 19]
```diff
@@ -46,6 +45,7 @@
               mgt.state_transitions = '';
               mgt.notification_icons = LiveObject.icons(mgt);
             });
+            window.generateCommandDropdown.cleanupButtons(tableInstance[0]);
```
Comment (jgrund) [inline templates/mgt_tab.html]: When does this get called?

File: static/js/api.js (hunk 3 above)
Comment (johnsonw) [reply static/js/api.js]: done

File: static/js/services/generate_command_dropdown.js (hunk 5 above)
Comment (johnsonw) [reply static/js/services/generate_command_dropdown.js]: For sure. I left these in so you could see at what point they were being called in the demo.

File: static/js/services/generate_command_dropdown.js (hunk 7 above)
Comment (johnsonw) [reply static/js/services/generate_command_dropdown.js]: removed.

File: static/js/services/generate_command_dropdown.js (hunk 8 above)
Comment (johnsonw) [reply static/js/services/generate_command_dropdown.js]: removed.

File: static/js/services/generate_command_dropdown.js (hunk 9 above)
Comment (johnsonw) [reply static/js/services/generate_command_dropdown.js]: removed.

File: static/js/services/generate_command_dropdown.js (hunk 11 above)
Comment (johnsonw) [reply static/js/services/generate_command_dropdown.js]: removed.

File: static/js/services/generate_command_dropdown.js (hunk 12 above)
Comment (johnsonw) [reply static/js/services/generate_command_dropdown.js]: removed.

File: static/js/services/generate_command_dropdown.js (hunk 14 above)
Comment (johnsonw) [reply static/js/services/generate_command_dropdown.js]: It doesn't. It was initially in a different file so I needed window in order to access it. Since I moved it to this file I no longer need it.

File: static/js/services/generate_command_dropdown.js (hunk 15 above)
Comment (johnsonw) [reply static/js/services/generate_command_dropdown.js]: Same as above. Updated.

File: static/js/target_detail.js (hunk 16 above)
Comment (johnsonw) [reply static/js/target_detail.js]: This one is not added to a data table but is instead simply an element on the dom. When the page unload event occurs it will be cleaned up.

File: templates/base.html (hunk 17 above)
Comment (johnsonw) [reply templates/base.html]: I must have turned my formatter off after this file. I'll fix.

File: static/js/target_detail.js (hunk 16 above)
Comment (jgrund) [reply static/js/target_detail.js]: Isn't target_detail a modal though?

File: static/js/api.js (hunk 4 above)
Comment (johnsonw) [reply static/js/api.js]: updated.

File: static/js/services/generate_command_dropdown.js (hunk 6 above)
Comment (johnsonw) [reply static/js/services/generate_command_dropdown.js]: removed.

File: templates/mgt_tab.html (hunk 19 above)
Comment (johnsonw) [reply templates/mgt_tab.html]: Looks like this is dead code. I will remove.

File: static/js/services/generate_command_dropdown.js [hunk 20]
```diff
@@ -1,96 +1,123 @@
...
-            $scope.safeApply(function () { $scope.$destroy(); }, $scope);
+  addListener(uuid, locks => instance.set_locks(locks));
 
-            fragment.unbind('$destroy');
-          });
-        }, $scope);
-      }
-    };
-  }
+  addDestroyer(uuid,  () => {
```
Comment (jgrund) [inline static/js/services/generate_command_dropdown.js]: Extra space

Comment (johnsonw) [reply static/js/services/generate_command_dropdown.js]: fixed

File: iml-old-gui.spec [hunk 21]
```diff
@@ -1,8 +1,8 @@
-﻿%define base_name old-gui
+%define base_name old-gui
 
 Name:       iml-%{base_name}
-Version:    @version@
-Release:    @release@%{?dist}
+Version:    3.1.0
+Release:    1%{?dist}
```
Comment (jgrund) [inline iml-old-gui.spec]: Do we need the `#Release {Start|End}` wrapper here?

Comment (johnsonw) [reply iml-old-gui.spec]: Yes
